File: src/repotask/kb/schema.py
"""Knowledge base manifest, document frontmatter, and slice definitions."""

from __future__ import annotations

from pathlib import Path
from typing import Any, Literal

import yaml
from pydantic import BaseModel, ConfigDict, Field, ValidationError

from repotask.errors import RepoTaskError

KB_SCHEMA_VERSION = 1
MANIFEST_NAME = "kb.yaml"

Layer = Literal["convention", "recipe"]

FRONTMATTER_FENCE = "---"


class KbModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class FactFamily(KbModel):
    """A curated project-fact family the indexer knows how to populate."""

    name: str
    description: str = ""
    stacks: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    kinds: list[str] = Field(default_factory=list)
    name_pattern: str = ""
    path_pattern: str = ""


class Manifest(KbModel):
    schema_version: int
    name: str = "knowledge-base"
    description: str = ""
    default_budget: int = 6000
    fact_families: list[FactFamily] = Field(default_factory=list)

    def family(self, name: str) -> FactFamily | None:
        return next((item for item in self.fact_families if item.name == name), None)


class Slice(KbModel):
    """Which knowledge applies to a stack, optionally narrowed per intent."""

    stack: str
    extends: list[str] = Field(default_factory=list)
    conventions: list[str] = Field(default_factory=list)
    recipes: list[str] = Field(default_factory=list)
    facts: list[str] = Field(default_factory=list)
    intents: dict[str, list[str]] = Field(default_factory=dict)


class Document(KbModel):
    """A convention or recipe markdown file with parsed frontmatter."""

    id: str
    title: str
    layer: Layer
    stacks: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    applies_to: list[str] = Field(default_factory=list)
    budget_hint: int = 0
    body: str = ""
    path: str = ""


def parse_frontmatter(text: str) -> tuple[dict[str, Any], str]:
    """Split `---` delimited YAML frontmatter from the markdown body."""
    lines = text.splitlines()
    if not lines or lines[0].strip() != FRONTMATTER_FENCE:
        return {}, text
    for index in range(1, len(lines)):
        if lines[index].strip() == FRONTMATTER_FENCE:
            raw = yaml.safe_load("\n".join(lines[1:index])) or {}
            if not isinstance(raw, dict):
                raise RepoTaskError("Document frontmatter must be a YAML mapping.")
            return raw, "\n".join(lines[index + 1 :]).strip()
    raise RepoTaskError("Document frontmatter is missing its closing `---`.")


def load_document(path: Path, layer: Layer, kb_root: Path) -> Document:
    front, body = parse_frontmatter(path.read_text(encoding="utf-8"))
    front.setdefault("id", path.stem)
    front.setdefault("title", path.stem.replace("-", " ").title())
    # The directory decides the layer. Frontmatter may restate it, but not contradict it.
    declared = front.pop("layer", layer)
    if declared != layer:
        raise RepoTaskError(
            f"{path} declares layer '{declared}' but sits in the {layer} directory."
        )
    try:
        return Document(
            **front, layer=layer, body=body, path=str(path.relative_to(kb_root))
        )
    except ValidationError as error:
        raise RepoTaskError(f"Invalid frontmatter in {path}: {_first_message(error)}") from error


def load_manifest(path: Path) -> Manifest:
    raw = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    if not isinstance(raw, dict):
        raise RepoTaskError(f"{path} must contain a YAML mapping.")
    try:
        manifest = Manifest(**raw)
    except ValidationError as error:
        raise RepoTaskError(f"Invalid {MANIFEST_NAME}: {_first_message(error)}") from error
    if manifest.schema_version != KB_SCHEMA_VERSION:
        raise RepoTaskError(
            f"Knowledge base schema_version {manifest.schema_version} is not supported; "
            f"this build reads version {KB_SCHEMA_VERSION}."
        )
    return manifest


def load_slice(path: Path) -> Slice:
    raw = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    if not isinstance(raw, dict):
        raise RepoTaskError(f"{path} must contain a YAML mapping.")
    raw.setdefault("stack", path.stem)
    try:
        return Slice(**raw)
    except ValidationError as error:
        raise RepoTaskError(f"Invalid slice {path}: {_first_message(error)}") from error


def _first_message(error: ValidationError) -> str:
    item = error.errors()[0]
    location = ".".join(str(part) for part in item["loc"]) or "<root>"
    return f"{location}: {item['msg']}"

File: src/kb/schema.rs
//! Knowledge base manifest, document frontmatter, and slice definitions.

use std::collections::BTreeMap;
use std::path::Path;

use anyhow::{bail, Context, Result};
use serde::{Deserialize, Serialize};

pub const KB_SCHEMA_VERSION: u32 = 1;
pub const MANIFEST_NAME: &str = "kb.yaml";
const FRONTMATTER_FENCE: &str = "---";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Layer {
    Convention,
    Recipe,
}

impl Layer {
    pub fn as_str(self) -> &'static str {
        match self {
            Layer::Convention => "convention",
            Layer::Recipe => "recipe",
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FactFamily {
    pub name: String,
    #[serde(default)]
    pub description: String,
    /// Project stacks this family applies to. Empty means every stack. A family
    /// that matches none of `project.stacks` is not built at all.
    #[serde(default)]
    pub stacks: Vec<String>,
    /// Languages a symbol may be written in. Empty means every language — prefer
    /// `path_pattern` over listing languages to keep a family off other stacks.
    #[serde(default)]
    pub languages: Vec<String>,
    #[serde(default)]
    pub kinds: Vec<String>,
    #[serde(default)]
    pub name_pattern: String,
    #[serde(default)]
    pub path_pattern: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Manifest {
    pub schema_version: u32,
    #[serde(default = "default_kb_name")]
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default = "default_kb_budget")]
    pub default_budget: usize,
    #[serde(default)]
    pub fact_families: Vec<FactFamily>,
}

fn default_kb_name() -> String {
    "knowledge-base".into()
}

fn default_kb_budget() -> usize {
    6000
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Slice {
    #[serde(default)]
    pub stack: String,
    #[serde(default)]
    pub extends: Vec<String>,
    #[serde(default)]
    pub conventions: Vec<String>,
    #[serde(default)]
    pub recipes: Vec<String>,
    #[serde(default)]
    pub facts: Vec<String>,
    #[serde(default)]
    pub intents: BTreeMap<String, Vec<String>>,
}

/// Frontmatter as written on disk. `layer` may restate the directory, never contradict it.
#[derive(Debug, Default, Deserialize)]
#[serde(deny_unknown_fields)]
struct Frontmatter {
    #[serde(default)]
    id: Option<String>,
    #[serde(default)]
    title: Option<String>,
    #[serde(default)]
    layer: Option<Layer>,
    #[serde(default)]
    stacks: Vec<String>,
    #[serde(default)]
    tags: Vec<String>,
    #[serde(default)]
    applies_to: Vec<String>,
    #[serde(default)]
    budget_hint: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct Document {
    pub id: String,
    pub title: String,
    pub layer: Layer,
    pub stacks: Vec<String>,
    pub tags: Vec<String>,
    pub applies_to: Vec<String>,
    pub budget_hint: usize,
    pub body: String,
    pub path: String,
}

/// Split `---` delimited YAML frontmatter from the markdown body.
pub fn parse_frontmatter(text: &str) -> Result<(String, String)> {
    let lines: Vec<&str> = text.lines().collect();
    if lines.first().map(|line| line.trim()) != Some(FRONTMATTER_FENCE) {
        return Ok((String::new(), text.to_string()));
    }
    for (index, line) in lines.iter().enumerate().skip(1) {
        if line.trim() == FRONTMATTER_FENCE {
            let front = lines[1..index].join("\n");
            let body = lines[index + 1..].join("\n").trim().to_string();
            return Ok((front, body));
        }
    }
    bail!("Document frontmatter is missing its closing `---`.");
}

pub fn load_document(path: &Path, layer: Layer, kb_root: &Path) -> Result<Document> {
    let text = std::fs::read_to_string(path)
        .with_context(|| format!("Could not read {}", path.display()))?;
    let (front_text, body) = parse_frontmatter(&text)?;
    let front: Frontmatter = if front_text.trim().is_empty() {
        Frontmatter::default()
    } else {
        serde_norway::from_str(&front_text)
            .with_context(|| format!("Invalid frontmatter in {}", path.display()))?
    };

    if let Some(declared) = front.layer {
        if declared != layer {
            bail!(
                "{} declares layer '{}' but sits in the {} directory.",
                path.display(),
                declared.as_str(),
                layer.as_str()
            );
        }
    }

    let stem = path
        .file_stem()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();
    let relative = path
        .strip_prefix(kb_root)
        .unwrap_or(path)
        .to_string_lossy()
        .replace('\\', "/");
    Ok(Document {
        id: front.id.unwrap_or_else(|| stem.clone()),
        title: front.title.unwrap_or_else(|| title_case(&stem)),
        layer,
        stacks: front.stacks,
        tags: front.tags,
        applies_to: front.applies_to,
        budget_hint: front.budget_hint,
        body,
        path: relative,
    })
}

fn title_case(stem: &str) -> String {
    stem.split('-')
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

pub fn load_manifest(path: &Path) -> Result<Manifest> {
    let text = std::fs::read_to_string(path)
        .with_context(|| format!("Could not read {}", path.display()))?;
    let manifest: Manifest =
        serde_norway::from_str(&text).with_context(|| format!("Invalid {MANIFEST_NAME}"))?;
    if manifest.schema_version != KB_SCHEMA_VERSION {
        bail!(
            "Knowledge base schema_version {} is not supported; this build reads version {}.",
            manifest.schema_version,
            KB_SCHEMA_VERSION
        );
    }
    Ok(manifest)
}

pub fn load_slice(path: &Path) -> Result<Slice> {
    let text = std::fs::read_to_string(path)
        .with_context(|| format!("Could not read {}", path.display()))?;
    let mut slice: Slice = serde_norway::from_str(&text)
        .with_context(|| format!("Invalid slice {}", path.display()))?;
    if slice.stack.is_empty() {
        slice.stack = path
            .file_stem()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();
    }
    Ok(slice)
}
